hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,270 @@
+
+  private final int GZIP_HEADER_LEN = 10;
+  private final int GZIP_TRAILER_LEN = 8;
+
+  private Deflater deflater;
+
+  private int headerOff = 0;
+  private int trailerOff = 0;
+
+  private byte[] userBuf = null;
+  private int userBufOff = 0;
+  private int userBufLen = 0;
+
+  private int headerBytesWritten = 0;
+  private int trailerBytesWritten = 0;
+
+  private int currentBufLen = 0;
+
+  private final Checksum crc = DataChecksum.newCrc32();
+
+  private BuiltInGzipDecompressor.GzipStateLabel state;
+
+  public BuiltInGzipCompressor(Configuration conf) { init(conf); }
+
+  @Override
+  public boolean finished() {
+    // Only if the trailer is also written, it is thought as finished.
+    return deflater.finished() && state == BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+  }
+
+  @Override
+  public boolean needsInput() {
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      return deflater.needsInput();
+    }
+
+    return (state != BuiltInGzipDecompressor.GzipStateLabel.FINISHED);
+  }
+
+  @Override
+  public int compress(byte[] b, int off, int len) throws IOException {
+    int numAvailBytes = 0;
+
+    // If we are not within uncompressed data yet, output the header.
+    if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      if (userBufLen <= 0) {
+        return numAvailBytes;
+      }
+
+      int outputHeaderSize = writeHeader(b, off, len);
+      headerBytesWritten += outputHeaderSize;
+
+      // Completes header output.
+      if (headerOff == GZIP_HEADER_LEN) {
+        state = BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
+      }
+
+      numAvailBytes += outputHeaderSize;
+
+      if (outputHeaderSize == len) {
+        return numAvailBytes;
+      }
+
+      off += outputHeaderSize;
+      len -= outputHeaderSize;
+    }
+
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      // hand off user data (or what's left of it) to Deflater--but note that
+      // Deflater may not have consumed all of previous bufferload, in which case
+      // userBufLen will be zero
+      if (userBufLen > 0) {
+        deflater.setInput(userBuf, userBufOff, userBufLen);
+
+        crc.update(userBuf, userBufOff, userBufLen);  // CRC-32 is on uncompressed data
+
+        currentBufLen = userBufLen;
+        userBufOff += userBufLen;
+        userBufLen = 0;
+      }
+
+
+      // now compress it into b[]
+      int deflated = deflater.deflate(b, off, len);
+
+      numAvailBytes += deflated;
+      off += deflated;
+      len -= deflated;
+
+      // All current input are processed. Going to output trailer.
+      if (deflater.finished()) {
+        state = BuiltInGzipDecompressor.GzipStateLabel.TRAILER_CRC;
+        fillTrailer();
+      } else {
+        return numAvailBytes;
+      }
+    }
+
+    numAvailBytes += writeTrailer(b, off, len);
+    trailerBytesWritten += numAvailBytes;
+
+    if (trailerOff == GZIP_TRAILER_LEN) {
```

Comment:
nit: ditto - maybe move this into `writeTrailer` to keep it closer to its context.